File: hackerrank/ctci-is-binary-search-tree.java
class A {

    boolean checkBST(Node root) {
        if (root == null)
            return false;
        return checkBSTRange(root, null, null);
    }

    boolean checkBSTRange(Node root, Integer min, Integer max) {
        int data = root.data;
        Node right = root.right;
        Node left = root.left;
        boolean isInRange = satisfyRange(data, min, max);
        boolean rightIsGood = right != null && right.data > data && checkBSTRange(right, data, max) || right == null;
        boolean leftIsGood = left != null && left.data < data && checkBSTRange(left, min, data) || left == null;
        return isInRange && rightIsGood && leftIsGood;

    }

    boolean satisfyRange(int val, Integer min, Integer max) {
        boolean passMin = min != null && val > min || min == null;
        boolean passMax = max != null && val < max || max == null;
        return passMin && passMax;
    }

}

File: hackerrank/ctci-is-binary-search-tree.ts
type NodeT = {
  data: number;
  left?: NodeT;
  right?: NodeT;
};

function checkBST(rootNode?: NodeT, min?: number, max?: number): boolean {
  if (!rootNode) return false;

  const { data, right, left } = rootNode;
  const isInRange = satisfyRange(data, min, max);
  const rightIsGood =
    (right && right.data > data && checkBST(right, data, max)) || !right;
  const leftIsGood =
    (left && left.data < data && checkBST(left, min, data)) || !left;
  return isInRange && !!rightIsGood && !!leftIsGood;
}

const satisfyRange = (val: number, min?: number, max?: number): boolean =>
  ((min && val > min) || !min) && ((max && val < max) || !max);

const root: NodeT = {
  data: 4,
  left: {
    data: 2,
    left: { data: 1 },
    right: { data: 3 }
  },
  right: {
    data: 6,
    left: { data: 5 },
    right: { data: 7 }
  }
};

// console.log(checkBST(root, undefined, undefined));
